### ISSUE TITLE
Vim integration changes changelist and marks

### ISSUE THREAD

**Issue by smr97** [psf/black#1001]
Operating system:
OSX - Mojave
Python version:
3.6.7
Black, version 19.3b0 on Python 3.7.4 (default, Jul  9 2019, 18:13:23)
[Clang 10.0.1 (clang-1001.0.46.4)].
Does also happen on master:
didn't check, but probably.

Hey, I love using black and have never had issues with formatting or generated code. Thanks for this plugin. I use the vim integration of black, installed using vim-plug (the line reads "Plug 'ambv/black' ''.

The issue I am having is that whenever black formats the file and modifies the buffer (?), my changelist and my marks are lost. My jumplist last item invariably takes me to the top of the file. Is there some way I can stop this from happening? I am kind of a noob to vim, so please excuse me if this is just not possible....

**Comment by cwinsnes** [comment]
Could the reason for this happening be that Black is overwriting the entire buffer with the line `vim.current.buffer[:] = new_buffer_str.split('\n')[:-1]`? If you delete a line with a mark, the mark disappears with it. Similarly, the last edit would now become an insert from position 0 in the file, so the jumplist would go to the first line.

**Comment by smr97** [comment]
Yes, I saw this line and was wondering the same thing. Do you have an
alternate solution though? I dont know how to fix this logically.

**Comment by cwinsnes** [comment]
In vimscript there are the functions `marks` and `getjumplist`. Maybe we could store those as temporary variables before calling Black and then restore them afterwards? It's not going to be perfect, as it won't take the line changes by Black into account.

**Comment by smr97** [comment]
Interesting, I don't know much of vimscript to be honest, but this might be a decent hack. However, how do we set the marks in the new buffer to the value of this temporary variable (after the file is formatted and buffer contents are swapped)?

The larger issue would also be, is it possible to not overwrite the buffer with formatted contents, but kind of diff it and update the diffs only. Again, I don't know how to do this in vimscript.

I can take a look this weekend, but if you beat me to it, do let me know and I'll try out your branch with the changes!

Thanks!

**Comment by gegnew** [comment]
Anyone do anything with this?
